flow360/component/simulation/web/asset_base.py

```diff
@@ -0,0 +1,172 @@
+"""Base class of resource/asset (geometry, surface mesh, volume mesh and case)"""
+
+from __future__ import annotations
+
+import json
+import time
+from abc import ABCMeta, abstractmethod
+from datetime import datetime
+from typing import List, Union
+
+from flow360.cloud.requests import LengthUnitType
+from flow360.cloud.rest_api import RestApi
+from flow360.component.interfaces import BaseInterface, DraftInterface, ProjectInterface
+from flow360.component.resource_base import (
+    AssetMetaBaseModel,
+    Flow360Resource,
+    ResourceDraft,
+)
+from flow360.component.simulation.simulation_params import SimulationParams
+from flow360.component.utils import validate_type
+from flow360.log import log
+
+TIMEOUT_MINUTES = 60
+
+
+def _check_project_path_status(project_id: str, item_id: str, item_type: str) -> None:
+    RestApi(ProjectInterface.endpoint, id=project_id).get(
+        method="path", params={"itemId": item_id, "itemType": item_type}
+    )
+    # pylint: disable=fixme
+    # TODO: check all status on the given path
+
+
+class AssetBase(metaclass=ABCMeta):
+    """Base class for resource asset"""
+
+    _interface: type[BaseInterface] = None
+    _meta_class: type[AssetMetaBaseModel] = None
+    _draft_class: type[ResourceDraft] = None
+
+    @abstractmethod
+    def _retrieve_metadata(self) -> None:
+        # get the metadata when initializing the object (blocking)
+        pass
+
+    # pylint: disable=redefined-builtin
+    def __init__(self, id: str):
+        self._web = Flow360Resource(
+            interface=self._interface,
+            meta_class=self._meta_class,
+            id=id,
+        )
+        self._retrieve_metadata()
+        # get the project id according to resource id
+        resp = RestApi(self._interface.endpoint, id=id).get()
+        project_id = resp["projectId"]
+        solver_version = resp["solverVersion"]
+        self.project_id = project_id
+        self.solver_version = solver_version
+
+    @classmethod
+    # pylint: disable=protected-access
+    def _from_meta(cls, meta: AssetMetaBaseModel):
+        validate_type(meta, "meta", cls._meta_class)
+        resource = cls(id=meta.id)
+        resource._web._set_meta(meta)
+        return resource
+
+    @property
+    def info(self) -> AssetMetaBaseModel:
+        """Return the metadata of the resource"""
+        return self._web.info
+
+    @classmethod
+    def _interface(cls):
+        return cls._interface
+
+    @classmethod
+    def _meta_class(cls):
+        return cls._meta_class
```

Comment:
isn't it going to overshadow?

Reply:
What do you mean by overshadow?

Reply:
method `_interface()` has the same name as member `_interface`

Reply:
Good catch. Fixed